src/prefect/runner/_observers.py

```diff
@@ -0,0 +1,52 @@
+import asyncio
+import uuid
+from contextlib import AsyncExitStack
+from typing import Any, Protocol
+
+from prefect.events.clients import PrefectEventSubscriber, get_events_subscriber
+from prefect.events.filters import EventFilter, EventNameFilter
+from prefect.logging.loggers import get_logger
+
+
+class OnCancellingCallback(Protocol):
+    def __call__(self, flow_run_id: uuid.UUID) -> None: ...
+
+
+class FlowRunCancellingObserver:
+    def __init__(self, on_cancelling: OnCancellingCallback):
+        self.logger = get_logger("FlowRunCancellingObserver")
+        self.on_cancelling = on_cancelling
+        self._events_subscriber: PrefectEventSubscriber | None
+        self._exit_stack = AsyncExitStack()
+
+    async def _consume_events(self):
+        if self._events_subscriber is None:
+            raise RuntimeError(
+                "Events subscriber not initialized. Please use `async with` to initialize the observer."
+            )
+        async for event in self._events_subscriber:
+            flow_run_id = uuid.UUID(
+                event.resource["prefect.resource.id"].replace("prefect.flow-run.", "")
```

Comment:
weird and unlikely but technically i could call `emit_event` with an `event` named `prefect.flow-run.Cancelling` but have a resource id like `foo` and then it would say invalid UUID here right? is that worth catching here?

Reply:
Yeah, that's a good call. I can add handling for that.